### PR TITLE
Contribution guide for java

### PR DESCRIPTION
**Pull request by pavolloffay**
Contribution guide for java libraries https://github.com/opentracing-contrib/meta/issues/6.

@wu-sheng and @bensigelman could you please review? See also https://github.com/opentracing-contrib/meta/issues/6#issuecomment-273776751.

Think it would be better to move `Release Rights` to ot-java/RELEASE.md.

### CODE REVIEW

File: guide_java.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+```java
+   <dependency>
+       <groupId>io.opentracing.contrib</groupId>
+       <artifactId>opentracing-XXX</artifactId>
+   </dependency>
+```
+
+## Release
+General release process is described in 
+[opentracing-java/RELEASE.md](https://github.com/opentracing/opentracing-java/blob/master/RELEASE.md).
```
Comment (wu-sheng) [inline guide_java.md]: I think this release guide should describe like `opentracing-java` in literal, but should not ref it. `opentracing-java` should be released only by opentracing-team, but other contributors have their own repo in `opentracing-contrib`.

File: guide_java.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+Release is done by travis when pushing appropriate tag to the upstream repository. Configuration is done in
+[opentracing-java/pom.xml](https://github.com/opentracing/opentracing-java/blob/master/pom.xml). It depends on
+on `.settings.xml`, `travis.yml` and `travis/publish.sh`.
```
Comment (wu-sheng) [inline guide_java.md]: As a guide, I think we need provide samples about adding which codes into these three files.

File: guide_java.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+## Release Rights
```
Comment (wu-sheng) [inline guide_java.md]: ref comments on https://github.com/opentracing-contrib/meta/issues/6#issuecomment-273933795.

Discuss there. 😊

File: guide_java.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Java Contribution Guide
+
+Contributors who would like host their java libraries in opentracing-contrib organization 
+should abide following rules:
+
+## Choosing ArtifactId and GroupId
+```java
+   <dependency>
+       <groupId>io.opentracing.contrib</groupId>
+       <artifactId>opentracing-XXX</artifactId>
```
Comment (pavolloffay) [inline guide_java.md]: Why?

Currently we have:

`opentracing-jaxrs2`
`opentracing-web-servlet-filter`
`opentracing-globaltracer` https://github.com/opentracing-contrib/java-globaltracer/blob/master/pom.xml#L24

How other is it done in other frameworks? https://mvnrepository.com/artifact/org.springframework/spring-context/4.3.5.RELEASE

File: guide_java.md (hunk 1 above)
Comment (pavolloffay) [inline guide_java.md]: As repos in  `opentracing-contrib`  use the same release process It's better to have it in one place. It's easier to maintain/manage.

File: guide_java.md (hunk 2 above)
Comment (pavolloffay) [inline guide_java.md]: partially agree but code might vary in some cases. It will have to be reviewed carefully.

File: guide_java.md (hunk 4 above)
Comment (wu-sheng) [inline guide_java.md]: Ignore this, my mistake.

Comment (bensigelman) [inline guide_java.md]: @pavolloffay it would be helpful if we could establish a preference for the `XXX` in terms of the github repository name. E.g., if the repo is `java-spanmanager`, the artifact is `opentracing-spanmanager` (or similar).

File: guide_java.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+   </dependency>
+```
+
+## Release
+General release process is described in 
+[opentracing-java/RELEASE.md](https://github.com/opentracing/opentracing-java/blob/master/RELEASE.md).
+
+Release is done by travis when pushing appropriate tag to the upstream repository. Configuration is done in
```
Comment (bensigelman) [inline guide_java.md]: IIUC there is also an expectation that a `mvnw` script resides in the repo... is that going to be obvious to any reader?

File: guide_java.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+[opentracing-java/pom.xml](https://github.com/opentracing/opentracing-java/blob/master/pom.xml). It depends on
+on `.settings.xml`, `travis.yml` and `travis/publish.sh`.
+
+## Release Rights
+For the successful release several access right are needed:
```
Comment (bensigelman) [inline guide_java.md]: I'm fine with a pointer to `@bensigelman` in this doc for bintray approvals... their notification system is bad so I won't necessarily notice unless someone reaches out, and I don't like to block people.

File: guide_java.md (hunk 5 above)
Comment (wu-sheng) [inline guide_java.md]: Mvnw seems not necessary by the guides. @pavolloffay what is true? If need it to release, please add it to the guide.

File: guide_java.md (hunk 2 above)
Comment (wu-sheng) [inline guide_java.md]: As releasing a guide, we should have responsibilities to make sure, it is easy enough, even for a rookie. Particular this guide targets `opentracing-contrib`'s contributors, can not make sure any of us have enough times to answer questions about this.

**Comment by wu-sheng** [comment]
@pavolloffay , I will do step by step, by following this guide. And ask you to review the [branch](https://github.com/opentracing-contrib/java-hprose/tree/feature/first-release).

Let's see whether the guide is enough?

**Comment by wu-sheng** [comment]
Already send a pull request, https://github.com/opentracing-contrib/java-hprose/pull/3. Can you review it?

**Comment by pavolloffay** [comment]
@wu-sheng review done. Thanks for step by step replication :+1: . 

Couple of things need to me handled here:
1. package name in Bintray should not contain invalid characters. We cold do the convention to match repo name or artifact id. I more like artifact id, repo is linked in bintray package description.
2. release profile to attach javadocs and sources.

**Comment by pavolloffay** [comment]
@sjoerdtalsma could you please review too?

File: guide_java.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
+# Java Contribution Guide
```
Comment (bensigelman) [inline guide_java.md]: let's link to this guide from the main `opentracing-contrib/meta` README

File: guide_java.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+* XXX - stands for a repository name without java prefix 
+
+## Release configuration
+We encourage you to use the same release process as for `opentracing-api` artifact.
+General release process is described in 
+[opentracing-java/RELEASE.md](https://github.com/opentracing/opentracing-java/blob/master/RELEASE.md).
+
+Release is done by travis when pushing an appropriate tag to the upstream repository. Configuration is done in
```
Comment (sjoerdtalsma) [inline guide_java.md]: >when pushing an appropriate tag to the upstream repository.

Maybe provide an example e.g. `release-1.2.3`?

File: guide_java.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+[opentracing-java/pom.xml](https://github.com/opentracing/opentracing-java/blob/master/pom.xml). It depends on
+on `.settings.xml`, `travis.yml` and `travis/publish.sh`. Note that scripts are using maven wrapper, `./mvnw` to
```
Comment (sjoerdtalsma) [inline guide_java.md]: `travis.yml` -> `.travis.yml` (I think)

File: guide_java.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+We encourage you to use the same release process as for `opentracing-api` artifact.
+General release process is described in 
+[opentracing-java/RELEASE.md](https://github.com/opentracing/opentracing-java/blob/master/RELEASE.md).
+
+Release is done by travis when pushing an appropriate tag to the upstream repository. Configuration is done in
+[opentracing-java/pom.xml](https://github.com/opentracing/opentracing-java/blob/master/pom.xml). It depends on
+on `.settings.xml`, `travis.yml` and `travis/publish.sh`. Note that scripts are using maven wrapper, `./mvnw` to 
+pin a consistent version. 
+
+Wrapper can be generated by `mvn -N io.takari:maven:wrapper -Dmaven=3.3.9`.
```
Comment (sjoerdtalsma) [inline guide_java.md]: Make sure to add a `.gitignore` file containing `!maven-wrapper.jar` in the `.mvn/wrapper` directory after generation, to include the `maven-wrapper.jar` in git.

(I prefer this approach to including _all_ jars at repo level as it is more specific)

File: guide_java.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+ 1. [Bintray](https://bintray.com/opentracing/maven/opentracing-java) - register and request access directly in the organization, if delayed ping @bensigelman on our [gitter](https://gitter.im/opentracing/public) channel
+ 2. [Sonatype](https://issues.sonatype.org/browse/OSSRH-19982) - request access directly in the issue
+
+These need to be re-encrypted and added to `.travis.yml`.
+
+## Sync to JCenter (maven central)
+By default package is not published to maven central. To be synced it needs to be 
+included in `JCenter` repository. This has to be done manually in Bintray UI after 
+the first release. On following releases this is done automatically by `centralsync-maven-plugin`.
```
Comment (sjoerdtalsma) [inline guide_java.md]: typo:
> by `centralsync-maven-plugin`

->
> by _the_ `centralsync-maven-plugin`

File: guide_java.md (hunk 10 above)
Comment (pavolloffay) [inline guide_java.md]: I have added a comment that wrapper adds jar file to the repo.

I will leave the configuration of `.gitgnore` to the user.

Comment (wu-sheng) [inline guide_java.md]: Add a tip, remind readers to change .gitignore.

**Comment by pavolloffay** [comment]
PR updated:

* note about jar files ignored by `.gitignore`. 
* link to sonatype `pom.xml` requirements 
* reformatted README.md and add link to guide_java

**Comment by bensigelman** [comment]
@pavolloffay anything blocking merge on this? Seems like several JVM folks have taken a look at this point.

My only *possible* request would be to try (?) and get as much of the boilerplate as is possible submitted in its own PR. Sometimes the reviews can be annoying when there's a giant `mvnw` blob in the middle, etc.

File: guide_java.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+  </profiles>
+```
+
+* XXX - stands for maven package name in Bintray. It can be equal to the repository name or artifactId. We encourage you to create it in Bintray UI before the first release.
+
+Please submit this configuration in an individual pull request, to see that everything is properly configured.
```
Comment (pavolloffay) [inline guide_java.md]: > > My only possible request would be to try (?) and get as much of the boilerplate as is possible submitted in its own PR. 

@bensigelman I have added this. Please have a look if it is fine. Or we can split  boilerplate code differently.

**Comment by pavolloffay** [comment]
@sjoerdtalsma did you have to create maven packages in bintray before release? Yesterday I released java-okhttp and it failed on:  https://travis-ci.org/opentracing-contrib/java-okhttp/builds/199774438 
```
Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.7:deploy (default-deploy) on project opentracing-okhttp-parent: Failed to deploy artifacts: Could not find artifact io.opentracing.contrib:opentracing-okhttp-parent:pom:0.0.1 in bintray (https://api.bintray.com/maven/opentracing/maven/opentracing-okhttp/;publish=1) -> [Help 1]
``` 
then when I created package in bintray and deployed again it worked fine.

For now I have added a comment that users should create package in bintray before the first release.

File: guide_java.md [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+  </profiles>
+```
+
+* XXX - stands for maven package name in Bintray. It can be equal to the repository name or artifactId. We encourage you to create it in Bintray UI before the first release.
```
Comment (pavolloffay) [inline guide_java.md]: create package in bintray before the first release.

cc @bensigelman @sjoerdtalsma

**Comment by pavolloffay** [comment]
@bensigelman I think it is ready to go in.

**Comment by bensigelman** [comment]
@pavolloffay awesome, thank you so much!
